File: software/zyarm_sdk/python/src/zyarm_sdk/teleop.py
from __future__ import annotations

import threading
from typing import Optional

from .arm import ZyArm
from .config import TeleopConfig
from .errors import StaleStateError
from .protocol import MasterDataFrame
from .retarget import RetargetConfig, Retargeter
from .types import StateSource, TeleopAction, TeleopStepResult


class ZyArmLeader:
    def __init__(self, arm: ZyArm, config: Optional[TeleopConfig] = None) -> None:
        self.arm = arm
        self.config = config or TeleopConfig()

    def start(self) -> None:
        self.arm.enter_master_mode(frequency_hz=self.config.leader_hz)

    def stop(self) -> None:
        self.arm.stop_master_mode()

    def get_action(
        self,
        *,
        max_age_ms: Optional[float] = None,
        wait: bool = False,
        timeout_ms: Optional[float] = None,
    ) -> Optional[TeleopAction]:
        frame = None
        if wait:
            before = self.arm.transport.master_data_sequence
            frame = self.arm.transport.wait_for_master_data_after(
                before,
                (self.config.wait_timeout_ms if timeout_ms is None else timeout_ms) / 1000.0,
            )
        else:
            frame = self.arm.transport.latest_master_data()
        if frame is None:
            return None
        return self.action_from_frame(frame, max_age_ms=max_age_ms)

    def action_from_frame(
        self,
        frame: MasterDataFrame,
        *,
        max_age_ms: Optional[float] = None,
    ) -> TeleopAction:
        action = TeleopAction(
            positions=tuple(self.arm.mapping.hardware_to_public(frame.values)),
            source=StateSource.MASTER_DATA,
            timestamp=frame.received_at,
            sequence=frame.sequence,
            raw_line=frame.raw_line,
        )
        limit = self.config.action_max_age_ms if max_age_ms is None else max_age_ms
        if action.age_ms > limit:
            raise StaleStateError(f"Leader action is stale: {action.age_ms:.1f} ms > {limit:.1f} ms")
        return action


class ZyArmFollower:
    def __init__(
        self,
        arm: ZyArm,
        *,
        retarget: Optional[RetargetConfig] = None,
        config: Optional[TeleopConfig] = None,
    ) -> None:
        self.arm = arm
        self.config = config or TeleopConfig()
        self.retargeter = Retargeter(retarget)

    def send_action(self, action: TeleopAction, *, wait_state: bool = False) -> TeleopStepResult:
        positions = self.retargeter.apply(action)
        command = self.arm.fast_io(
            positions,
            self.retargeter.apply_mask,
            wait_state=wait_state,
            timeout_ms=self.config.wait_timeout_ms,
        )
        observation = self.arm.get_latest_state(self.config.state_max_age_ms)
        return TeleopStepResult(action=action, command=command, observation=observation)

    def get_observation(self):
        return self.arm.get_latest_state(self.config.state_max_age_ms)


class ZyArmTeleopPair:
    def __init__(
        self,
        leader: ZyArm,
        follower: ZyArm,
        *,
        config: Optional[TeleopConfig] = None,
        retarget: Optional[RetargetConfig] = None,
    ) -> None:
        self.config = config or TeleopConfig()
        self.leader = ZyArmLeader(leader, self.config)
        self.follower = ZyArmFollower(follower, retarget=retarget, config=self.config)
        self._follow_stop = threading.Event()
        self._follow_thread: Optional[threading.Thread] = None
        self._follower_slave_started = False

    def connect(self) -> "ZyArmTeleopPair":
        self.leader.arm.connect()
        self.follower.arm.connect()
        return self

    def start_step_mode(self) -> None:
        try:
            self._start_follower_slave_mode()
            self.leader.start()
        except Exception:
            self._stop_follower_slave_mode()
            raise

    def step(self, *, wait_action: bool = False, wait_state: bool = False) -> Optional[TeleopStepResult]:
        action = self.leader.get_action(wait=wait_action)
        if action is None:
            return None
        return self.follower.send_action(action, wait_state=wait_state)

    def start_auto_follow(self) -> None:
        if self._follow_thread and self._follow_thread.is_alive():
            return
        self._follow_stop.clear()
        self.start_step_mode()
        last_sequence = self.leader.arm.transport.master_data_sequence
        self._follow_thread = threading.Thread(
            target=self._auto_follow_loop,
            args=(last_sequence,),
            daemon=True,
        )
        self._follow_thread.start()

    def stop(self) -> None:
        self._follow_stop.set()
        if self._follow_thread and self._follow_thread.is_alive():
            self._follow_thread.join(timeout=1.0)
        self._follow_thread = None
        self.leader.stop()
        self._stop_follower_slave_mode()

    def close(self) -> None:
        self.stop()
        self.leader.arm.close()
        self.follower.arm.close()

    def _start_follower_slave_mode(self) -> None:
        if self._follower_slave_started:
            return
        result = self.follower.arm.set_master_slave_lpf(self.config.slave_filter_lpf_alpha)
        if not result.accepted:
            raise RuntimeError("follower slave filter LPF configuration failed")
        result = self.follower.arm.enter_slave_mode()
        if not result.accepted:
            raise RuntimeError("follower slave mode start failed")
        self._follower_slave_started = True

    def _stop_follower_slave_mode(self) -> None:
        if not self._follower_slave_started:
            return
        try:
            self.follower.arm.stop_master_mode()
        finally:
            self._follower_slave_started = False

    def _auto_follow_loop(self, last_sequence: int) -> None:
        while not self._follow_stop.is_set():
            frame = self.leader.arm.transport.wait_for_master_data_after(
                last_sequence,
                self.config.wait_timeout_ms / 1000.0,
            )
            if frame is None:
                continue
            last_sequence = frame.sequence
            try:
                action = self.leader.action_from_frame(frame)
                self.follower.send_action(action, wait_state=False)
            except StaleStateError:
                continue
            except Exception:
                continue

File: software/zyarm_sdk/cpp/src/teleop.cpp
#include "zyarm_sdk/teleop.hpp"

#include <chrono>
#include <thread>
#include <utility>

#include "zyarm_sdk/errors.hpp"

namespace zyarm_sdk
{

ZyArmLeader::ZyArmLeader(std::shared_ptr<ZyArm> arm, TeleopConfig config)
: arm_(std::move(arm)), config_(config)
{
}

void ZyArmLeader::start()
{
  arm_->enter_master_mode(config_.leader_hz);
}

void ZyArmLeader::stop()
{
  arm_->stop_master_mode();
}

std::optional<TeleopAction> ZyArmLeader::get_action(bool wait)
{
  std::optional<MasterDataFrame> frame;
  if (wait) {
    const auto before = arm_->transport()->master_data_sequence();
    frame = arm_->transport()->wait_for_master_data_after(before, config_.wait_timeout);
  } else {
    frame = arm_->transport()->latest_master_data();
  }
  if (!frame.has_value()) {
    return std::nullopt;
  }
  return action_from_frame(*frame);
}

TeleopAction ZyArmLeader::action_from_frame(const MasterDataFrame & frame) const
{
  TeleopAction action{
    arm_->mapping().hardware_to_public(frame.values),
    StateSource::MasterData,
    frame.received_at,
    frame.sequence,
    frame.raw_line};
  if (action.age_ms() > config_.action_max_age_ms) {
    throw StaleStateError("leader action is stale");
  }
  return action;
}

ZyArmFollower::ZyArmFollower(
  std::shared_ptr<ZyArm> arm,
  RetargetConfig retarget,
  TeleopConfig config)
: arm_(std::move(arm)), config_(config), retargeter_(retarget)
{
}

TeleopStepResult ZyArmFollower::send_action(const TeleopAction & action, bool wait_state)
{
  const auto positions = retargeter_.apply(action);
  auto command = arm_->fast_io(positions, retargeter_.apply_mask(), wait_state, config_.wait_timeout);
  auto observation = arm_->get_latest_state(config_.state_max_age_ms);
  return TeleopStepResult{action, command, observation};
}

std::optional<ArmState> ZyArmFollower::get_observation() const
{
  return arm_->get_latest_state(config_.state_max_age_ms);
}

ZyArmTeleopPair::ZyArmTeleopPair(
  std::shared_ptr<ZyArm> leader,
  std::shared_ptr<ZyArm> follower,
  TeleopConfig config,
  RetargetConfig retarget)
: config_(config),
  leader_arm_(leader),
  follower_arm_(follower),
  leader_(leader, config),
  follower_(follower, retarget, config)
{
}

ZyArmTeleopPair::~ZyArmTeleopPair()
{
  stop();
}

ZyArmTeleopPair & ZyArmTeleopPair::connect()
{
  leader_arm_->connect();
  follower_arm_->connect();
  return *this;
}

void ZyArmTeleopPair::start_step_mode()
{
  try {
    start_follower_slave_mode();
    leader_.start();
  } catch (...) {
    stop_follower_slave_mode();
    throw;
  }
}

std::optional<TeleopStepResult> ZyArmTeleopPair::step(bool wait_action, bool wait_state)
{
  auto action = leader_.get_action(wait_action);
  if (!action.has_value()) {
    return std::nullopt;
  }
  return follower_.send_action(*action, wait_state);
}

void ZyArmTeleopPair::start_auto_follow()
{
  if (follow_thread_.joinable()) {
    return;
  }
  stop_follow_ = false;
  start_step_mode();
  const auto last_sequence = leader_arm_->transport()->master_data_sequence();
  follow_thread_ = std::thread(&ZyArmTeleopPair::auto_follow_loop, this, last_sequence);
}

void ZyArmTeleopPair::stop()
{
  stop_follow_ = true;
  if (follow_thread_.joinable()) {
    follow_thread_.join();
  }
  leader_.stop();
  stop_follower_slave_mode();
}

void ZyArmTeleopPair::close()
{
  stop();
  leader_arm_->close();
  follower_arm_->close();
}

void ZyArmTeleopPair::auto_follow_loop(std::uint64_t last_sequence)
{
  while (!stop_follow_) {
    auto frame = leader_arm_->transport()->wait_for_master_data_after(
      last_sequence,
      config_.wait_timeout);
    if (!frame.has_value()) {
      continue;
    }
    last_sequence = frame->sequence;

    TeleopAction action;
    try {
      action = leader_.action_from_frame(*frame);
      follower_.send_action(action, false);
    } catch (...) {
      continue;
    }
  }
}

void ZyArmTeleopPair::start_follower_slave_mode()
{
  if (follower_slave_started_) {
    return;
  }
  auto result = follower_arm_->set_master_slave_lpf(config_.slave_filter_lpf_alpha);
  if (!result.accepted) {
    throw ZyArmError("follower slave filter LPF configuration failed");
  }
  result = follower_arm_->enter_slave_mode();
  if (!result.accepted) {
    throw ZyArmError("follower slave mode start failed");
  }
  follower_slave_started_ = true;
}

void ZyArmTeleopPair::stop_follower_slave_mode()
{
  if (!follower_slave_started_) {
    return;
  }
  follower_arm_->stop_master_mode();
  follower_slave_started_ = false;
}

}  // namespace zyarm_sdk
